### ISSUE TITLE
f2py/numpy-0.9.5 segfault (migrated from Trac #14)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#162]
Original ticket http://projects.scipy.org/numpy/ticket/14
Reported 2006-02-24 by unknown, assigned to @thouis not really:pearu.

I wrapped a fortran program with the f2py which comes with numpy-0.9.5 and
I get a segfault on import. To create the wrapper I did:
    f2py -c -m pythia pythia6326.f skip: pygaus pygau2
This successfully creates pythia.so. As soon as I import pythia.so I get the segfault:
    $ python
    Python 2.4.2 (#2, Nov 20 2005, 17:04:48)
    [GCC 4.0.3 20051111 (prerelease) (Debian 4.0.2-4)] on linux2
    Type "help", "copyright", "credits" or "license" for more information.
    >>> import pythia
    Segmentation fault
    $  
I get the same behaviour on 32 and 64 bit AMD processors (I have not tried Intel).

The fortran program is here: http://www.thep.lu.se/~torbjorn/pythia/pythia6326.f

**Comment by thouis** [comment]
Comment in Trac by @XXXthouis not really:pearu, 2006-02-24

Note to myself: this segfault is caused from wrapping character arrays in common blocks.
Segfault disappers when pyint6 and pydat4 common blocks are not wrapped.

**Comment by thouis** [comment]
Comment in Trac by @XXXthouis not really:teoliphant, 2006-03-08

The segfault is due to PyFortranObject_New returning NULL and this not being handled in PyDict_SetItemString.  

PyFortranObject_New is returning NULL because PyArray_New for variable-length arrays (i.e. character arrays) needs an itemsize specified (which is always 0) or else it fails.  This failure is not being caught.

So, there are two problems:

1) Setting the result of PyFortranObject_New should be handled if there is an error. 

   I added a new F2PyDict_SetItemString which special-cases obj=NULL to fix this problem.

2) If PyArray_STRING arrays are used they should be handled properly

   I fixed this so that the length of the string is the last dimension and the rank of the returned array is 1 smaller.  This seems to match what f2py is producing from the Fortran source file.
